crates/circuit/src/dag_circuit.rs

```diff
@@ -3319,77 +3286,121 @@ def _format(operand):
     ///
     /// A serial layer is a circuit with one gate. The layers have the
     /// same structure as in layers().
-    fn serial_layers(&self) -> PyResult<Py<PyIterator>> {
-        // for next_node in self.topological_op_nodes():
-        //     new_layer = self.copy_empty_like()
-        //
-        //     # Save the support of the operation we add to the layer
-        //     support_list = []
-        //     # Operation data
-        //     op = copy.copy(next_node.op)
-        //     qargs = copy.copy(next_node.qargs)
-        //     cargs = copy.copy(next_node.cargs)
-        //
-        //     # Add node to new_layer
-        //     new_layer.apply_operation_back(op, qargs, cargs, check=False)
-        //     # Add operation to partition
-        //     if not getattr(next_node.op, "_directive", False):
-        //         support_list.append(list(qargs))
-        //     l_dict = {"graph": new_layer, "partition": support_list}
-        //     yield l_dict
-        todo!()
+    fn serial_layers(&self, py: Python) -> PyResult<Py<PyIterator>> {
+        let layer_list = PyList::empty_bound(py);
+        for next_node in self.topological_op_nodes()? {
+            let retrieved_node: &PackedInstruction = match self.dag.node_weight(next_node) {
+                Some(NodeType::Operation(node)) => node,
+                _ => unreachable!("A non-operation node was obtained from topological_op_nodes."),
+            };
+            let mut new_layer = self.copy_empty_like(py)?;
+
+            // Save the support of the operation we add to the layer
+            let support_list = PyList::empty_bound(py);
+            let qubits = PyTuple::new_bound(
+                py,
+                self.qargs_cache
+                    .intern(retrieved_node.qubits_id)
+                    .iter()
+                    .map(|qubit| self.qubits.get(*qubit)),
+            )
+            .unbind();
+            new_layer.push_back(py, retrieved_node.clone())?;
+
+            if !retrieved_node.op.directive() {
+                support_list.append(qubits)?;
+            }
+
+            let layer_dict = [
+                ("graph", new_layer.into_py(py)),
+                ("partition", support_list.into_any().unbind()),
+            ]
+            .into_py_dict_bound(py);
+            layer_list.append(layer_dict)?;
+        }
+
+        Ok(layer_list.into_any().iter()?.into())
     }
 
     /// Yield layers of the multigraph.
-    fn multigraph_layers(&self) -> PyResult<Py<PyIterator>> {
-        // first_layer = [x._node_id for x in self.input_map.values()]
-        // return iter(rx.layers(self._multi_graph, first_layer))
-        todo!()
+    #[pyo3(name = "multigraph_layers")]
+    fn py_multigraph_layers(&self, py: Python) -> PyResult<Py<PyIterator>> {
+        let graph_layers = self.multigraph_layers().map(|layer| -> Vec<PyObject> {
+            layer
+                .into_iter()
+                .filter_map(|index| self.get_node(py, index).ok())
```

Comment:
Better to raise an error or at least panic if something weird like the node not being found happens, I'd say!